src/Console/CommandRunner.php

```diff
@@ -159,10 +161,39 @@ public function run(array $argv, ConsoleIo $io = null)
      * @param \Cake\Console\ConsoleIo $io The IO wrapper for the created shell class.
      * @param \Cake\Console\CommandCollection $commands The command collection to find the shell in.
      * @param string $name The command name to find
-     * @return \Cake\Console\Shell
+     * @return \Cake\Console\Shell|\Cake\Console\Command
      */
     protected function getShell(ConsoleIo $io, CommandCollection $commands, $name)
     {
+        $instance = $commands->get($name);
+        if (is_string($instance)) {
+            $instance = $this->createShell($instance, $io);
+        }
+        if (method_exists($instance, 'setRootName')) {
```

Comment:
Maybe an instance check would be better here?

```php
if ($instance instanceof Shell) {
     $instance->setRootName($this->root);
}
if ($instance instanceof Command) {
     $instance->setName("{$this->root} {$name}");
}
```

These are only two classes that the shell/command can extend.

Reply:
I can do that.